File: 21. Merge Two Sorted Lists/mergeTwoLists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

const arrayToListNode = (arr) => {
  const head = new ListNode();
  let list = head;
  arr.forEach((v) => {
    list.next = new ListNode(v);
    list = list.next;
  });
  return head.next;
};

const listNodeToArray = (listNode) => {
  const array = [];
  const recursive = (list) => {
    if (list === null) {
      return;
    }
    array.push(list.val);
    recursive(list.next);
  };
  recursive(listNode);
  return array;
};

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
const mergeTwoLists = (l1, l2) => {
  const list = new ListNode();
  let l = list;
  while (l1 !== null || l2 !== null) {
    if (l2 === null) {
      l.next = new ListNode(l1.val);
      l = l.next;
      l1 = l1.next;
    } else if (l1 === null) {
      l.next = new ListNode(l2.val);
      l = l.next;
      l2 = l2.next;
    } else if (l1.val < l2.val) {
      l.next = new ListNode(l1.val);
      l = l.next;
      l1 = l1.next;
    } else {
      l.next = new ListNode(l2.val);
      l = l.next;
      l2 = l2.next;
    }
  }
  return list.next;
};

console.log(
  listNodeToArray(
    mergeTwoLists(
      arrayToListNode([1, 2, 3, 4, 5, 7]),
      arrayToListNode([2, 3, 4, 5, 6])
    )
  )
);

File: leetcode/21. Merge Two Sorted Lists/mergeTwoLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

/**
 * @brief 12 m 50 s
 *
 */
class Solution {
 public:
  // list1 = [1, 2, 4]
  // list2 = [1, 3, 4]
  ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
    ListNode* dummy = new ListNode();  // dump = 0
    ListNode* tail = dummy;            // node = 0

    while (list1 != nullptr && list2 != nullptr) {
      if (list1->val <= list2->val) {
        tail->next = list1;  // 1. node = [0 1 2 4] 3. [1, 2, 4] 5. [3, 4]

        list1 = list1->next;  // 1. list1 = [2, 4] 3. [4] 5. []
      } else {
        tail->next = list2;  // 2. node = [1, 1, 3, 4] 4. [2, 3, 4]

        list2 = list2->next;  // 2. list2 = [3, 4] 4. [4]
      }

      tail = tail->next;  // 1. node = [1, 2, 4] 2. [1, 3, 4] 3. [2, 4] 4. [3,
                          // 4] 5. [4]
    }

    if (list1 != nullptr) {
      tail->next = list1;
    }

    if (list2 != nullptr) {
      tail->next = list2;  // [4, 4]
    }

    return dummy->next;
  }
};
